xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-default/src/main/java/org/xwiki/notifications/notifiers/internal/UserEventDispatcher.java

```diff
@@ -245,12 +245,12 @@ private void flush() throws Exception
                 for (Event event : it) {
                     try {
                         prefilterEvent(event, types);
+                        // We need to keep a list of handled events since the processing of the events is async:
+                        // without it, we might end up processing multiple times same events.
+                        handledEvents.add(event.getId());
                     } catch (Exception e) {
                         this.logger.warn("Failed to pre filter event with id [{}]: {}", event.getId(),
                             ExceptionUtils.getRootCauseMessage(e));
-
-                        // Remember the failed event to not query it again
-                        failedEvents.add(event.getId());
```

Comment:
How do you take care of this case now?

Reply:
I'm dropping this solution for something else, I haven't pushed yet since I'm still writing unit tests for it.